design-docs/trajectories.md

```diff
@@ -0,0 +1,188 @@
+for a differential drive robot, 
+which extends `Sample` and calculates the left and right wheel speeds
+given the chassis speeds and the drivetrain kinematics.
+
+### Trajectory
+
+The `Trajectory` class represents a trajectory,
+which will be implemented as a list of `Sample`s
+that are linearly interpolated between.
+
+The following methods will be implemented as 
+part of the public API of the `Trajectory` class:
+
+```java
+public class Trajectory {
+    public Time duration();
+    public Distance length();
+    
+    public Sample sampleAt(double time);
+    public Pose2d poseAt(double time);
+}
+```
+
+## Design: Trajectory Generator
+
+In addition to the main API,
+this API will also provide a trajectory generator.
+The trajectory generator will be used to generate trajectories
+for both differential drive and holonomic robots.
+
+It is intended to be used to generate trajectories that 
+are more complex than WPILib's previous trajectory API,
+but not to replace tools like Choreo and PathPlanner.
+
+### Builders
+
+The `TrajectoryBuilder` class is used to generate trajectories.
+It will be implemented as a builder pattern,
+with a fluent interface.
+
+```java
+Trajectory trajectory = TrajectoryBuilder.createWithConstraints(
+    startPose,
+    velConstraint, 
+    accelConstraint
+)
+.forward(distance)
+.turnBy(angle)
+.splineTo(waypoint1, tangent1)
+.build();
+```
+
+By default, the trajectory will use tangential heading interpolation.
+
+Other heading interpolation styles can be used instead
+on `lineTo` and `splineTo` methods:`
+- `linearHeading` linearly interpolates the heading
+- `splineHeading` interpolates the heading using a 1d spline
+- `constantHeading` uses the heading of the previous point
+
+```java
+Trajectory trajectory = TrajectoryBuilder.createWithConstraints(
+    startPose,
+    velConstraint, 
+    accelConstraint
+)
+.forward(distance)
+.turnBy(angle)
+.splineTo(waypoint1, tangent1)
+.splineToLinearHeading(waypoint2, heading1, tangent2)
+.lineToConstantHeading(waypoint3)
+.splineToSplineHeading(waypoint4, heading2, tangent3)
+.build();
+```
+
+The following paths will be supported:
+- `forward(Distance)` straight line in current direction
+- `turnBy(Angle)` turns desired angle
+- `turnTo(Rotation2d)`turns to reach desired heading
+- `splineTo(Translation2d, Rotation2d)` spline path to the waypoint with the given tangent
+- `lineTo(Translation2d)` straight line to given waypoint
+- `curveTo(Translation2d...)` Bézier curve path to the given waypoints
+
+### Constraints
+
+The `VelocityConstraint` and `AccelerationConstraint` classes
+will be used to define the constraints of the trajectory.
+
+WPILib currently contains a `TrajectoryConstraint` interface,
+which will be replaced by `VelocityConstraint` and `AccelerationConstraint`,
+and its implementations refactored to match the new interfaces.
+
+### Paths 
+
+The `PositionPath` interface will be used to represent 2D
```

Comment:
I'm torn on whether to include paths in addition to trajectories. I'm aware FTC uses path followers often, but FRC doesn't since time-based trajectory sampling and state feedback yield good performance already. I'd need to see a good use case presented not already covered by time-based trajectories.

We try to avoid offering multiple ways of doing the same thing since that means our docs have to explain the differences and tradeoffs. If the docs end up saying "just always use one thing unless you know what you're doing", that's grounds for removing the quasi-deprecated thing.